core/src/transport/incoming_limit.rs

```diff
@@ -0,0 +1,401 @@
+    ///
+    /// This is an optimization. This value must always be superior or equal
+    /// to [`IncomingLimitInner::current_num`]. If `current_num_approx` is
+    /// inferior to `limit`, then we know that `current_num` is also inferior
+    /// to `limit`.
+    current_num_approx: AtomicUsize,
+
+    /// Struct behind a mutex.
+    inner: Mutex<IncomingLimitInner>,
+}
+
+struct IncomingLimitInner {
+    /// Current number of connections being upgraded.
+    current_num: usize,
+
+    /// Collection of wakers filled if the limit is reached. Each listener
+    /// currently waiting for the limit may have an entry here, whose index
+    /// is [`IncomingLimitListen::waker_slot_num`].
+    wakers: Slab<Option<Waker>>,
+}
+
+impl IncomingLimit {
+    /// Creates a new [`IncomingLimit`] that applies the given limit.
+    pub fn new(limit: usize) -> Arc<Self> {
+        Arc::new(IncomingLimit {
+            limit,
+            current_num_approx: AtomicUsize::new(0),
+            inner: Mutex::new(IncomingLimitInner {
+                current_num: 0,
+                wakers: Slab::new(),
+            })
+        })
+    }
+}
+
+/// Applies the [`IncomingLimit`] to a transport.
+pub(crate) fn incoming_limit<T>(transport: T, limit: Arc<IncomingLimit>) -> IncomingLimitApply<T> {
+    IncomingLimitApply {
+        inner: transport,
+        limit,
+    }
+}
+
+/// Wraps around an implementation of [`Transport`] and applies the limit to
+/// the number of incoming connections.
+pub struct IncomingLimitApply<TInner> {
+    /// Underlying transport.
+    inner: TInner,
+    /// Limit system.
+    limit: Arc<IncomingLimit>,
+}
+
+impl<TInner> Transport for IncomingLimitApply<TInner>
+where
+    TInner: Transport,
+{
+    type Output = TInner::Output;
+    type Error = TInner::Error;
+    type Listener = IncomingLimitListen<TInner::Listener>;
+    type ListenerUpgrade = IncomingLimitUpgrade<TInner::ListenerUpgrade>;
+    type Dial = TInner::Dial;
+
+    fn listen_on(self, addr: Multiaddr) -> Result<Self::Listener, TransportError<Self::Error>> {
+        Ok(IncomingLimitListen {
+            inner: self.inner.listen_on(addr)?,
+            limit: self.limit.clone(),
+            waker_slot_num: None,
+        })
+    }
+
+    fn dial(self, addr: Multiaddr) -> Result<Self::Dial, TransportError<Self::Error>> {
+        self.inner.dial(addr)
+    }
+}
+
+impl<TInner> Clone for IncomingLimitApply<TInner>
+where
+    TInner: Clone
+{
+    fn clone(&self) -> Self {
+        IncomingLimitApply {
+            inner: self.inner.clone(),
+            limit: self.limit.clone(),
+        }
+    }
+}
+
+/// Reservation for a slot towards the limit.
+struct NumApproxGuard<'a> {
+    limit: &'a IncomingLimit,
+}
+
+impl<'a> NumApproxGuard<'a> {
+    fn acquire(limit: &'a IncomingLimit) -> (Self, usize) {
+        let current_num = limit.current_num_approx.fetch_add(1, Ordering::Acquire);
+        debug_assert_ne!(current_num, usize::max_value()); // Check for overflows.
```

Comment:
Debug builds already check for overflows. Why this additional check here?

Reply:
For atomics, the documentation explicitly says that it wraps around on overflow.